### PR TITLE
feat: auto col

### PR DESCRIPTION
**Pull request by lumixraku**
close https://github.com/dream-num/univer-pro/issues/2912

## Pull Request Checklist

- [ ] Related tickets or issues have been linked in the PR description (or missing issue).
- [ ] [Naming convention](https://github.com/dream-num/univer/blob/dev/docs/NAMING_CONVENTION.md) is followed (**do please** check it especially when you created new plugins, commands and resources).
- [ ] Unit tests have been added for the changes (if applicable).
- [ ] Breaking changes have been documented (or no breaking changes introduced in this PR).

### CODE REVIEW

File: packages/core/src/sheets/typedef.ts [hunk 1]
```diff
@@ -185,6 +185,16 @@ export interface IColumnData {
      */
     w?: number;
 
+    /**
+     * is current row self-adaptive to its content
+     */
+    ia?: BooleanNumber;
```
Comment (VicKun4937) [inline packages/core/src/sheets/typedef.ts:191]: remove

File: packages/docs-ui/src/controllers/render-controllers/doc.render-controller.ts [hunk 2]
```diff
@@ -76,8 +76,6 @@ export class DocRenderController extends RxDisposable implements IRenderModule {
             top: 0,
             bottom: 0,
             right: 0,
-            explicitViewportWidthSet: false,
```
Comment (VicKun4937) [inline packages/docs-ui/src/controllers/render-controllers/doc.render-controller.ts:79]: why  delete code in doc render

File: packages/engine-render/src/components/sheets/sheet-skeleton.ts [hunk 3]
```diff
@@ -636,6 +696,192 @@ export class SpreadsheetSkeleton extends Skeleton {
...
+            if (modelObject == null) {
+                return measuredWidth;
+            }
+
+            const { documentModel, textRotation } = modelObject;
+            if (documentModel == null) {
+                return measuredWidth;
+            }
+
+            // 策略： 不看 P 只看 V (效果不明显， 加上之前1.56s 之后 1.32s)
```
Comment (VicKun4937) [inline packages/engine-render/src/components/sheets/sheet-skeleton.ts:772]: English

File: packages/docs-ui/src/controllers/render-controllers/doc.render-controller.ts (hunk 2 above)
Comment (lumixraku) [reply packages/docs-ui/src/controllers/render-controllers/doc.render-controller.ts:79]: viewport do not need this config anymore

**Review by wzhudev — CHANGES_REQUESTED** [review]
Overall this PR looks great to me. Some code readability issue and misspelling can be fixed to improve code quality.

File: packages/engine-render/src/components/sheets/sheet-skeleton.ts [hunk 4]
```diff
@@ -636,6 +697,196 @@ export class SpreadsheetSkeleton extends Skeleton {
...
+        const visibleRangeViewMainTop = this.visibleRangeByViewportKey(SHEET_VIEWPORT_KEY.VIEW_MAIN_TOP);
+        if (visibleRangeViewMainTop) {
+            const { startRow: startRowOfViewMainTop, endRow: endRowOfViewMainTop } = visibleRangeViewMainTop;
+            for (let i = startRowOfViewMainTop; i <= endRowOfViewMainTop; i++) {
+                otherRowIndex.add(i);
+            }
+        }
+
+        for (const row of otherRowIndex) {
+            const { isMerged, isMergedMainCell } = this._getCellMergeInfo(colIndex, row);
```
[the remark this answers is not in the document]
Comment (lumixraku) [reply packages/engine-render/src/components/sheets/sheet-skeleton.ts:797]: a little repeat，but it's for viewMainTop col width.

**Comment by oumomomo** [comment]
1、输入超出单元格的内容使用自动换行，预期不会自动调整列宽仅换行
[2024_10_24 17_04_27 video.webm](https://github.com/user-attachments/assets/52d7d585-c1dd-4d38-b656-db57b073cdcf)

Google表现：
[2024_10_24 17_05_24 video.webm](https://github.com/user-attachments/assets/a27aacc4-2546-4bc9-9135-5652a747c422)

**Comment by oumomomo** [comment]
3、对有图标集的列使用自动调整列宽，预期展示图标集+内容
image

**Comment by oumomomo** [comment]
4、对自定义值的复选框使用自动调整列宽，预期调整的列宽预期跟复选框一致，而不是跟自定义值一致
[2024_10_24 19_39_51 video.webm](https://github.com/user-attachments/assets/294e9618-1493-41fd-b923-f8ba34ba85eb)
image


google表现：
image

**Comment by lumixraku** [comment]
> 3、对有图标集的列使用自动调整列宽，预期展示图标集+内容 [2024_10_24 19_26_46 video.webm](https://github.com/user-attachments/assets/9db27ec1-f72a-4265-ac7d-13d8c0964ab5) image

这是另一个问题， 排版的不对。 我设置了文字靠左， 都是 1234567 文字。 带有 icon 的把文字遮住了。 
自动列宽也是同样道理， 已经算了文字宽和列宽， 但是文字会被遮住。 是因为排版的时候没有考虑到 icon

**Comment by lumixraku** [comment]
> 2、自动调整列宽时内容未居中 image

我看了下 Google 并没有水平居中
设置列宽并不会导致文字水平对齐变化。而且通过调整水平对齐方式，也可以看到变化，表示列宽调整后并未居中。

https://github.com/user-attachments/assets/d9d32355-f7e0-464c-bc95-2fd1bd9f60e7

**Comment by lumixraku** [comment]
复选框没有复现


https://github.com/user-attachments/assets/978ac7a5-08b0-4925-9cb4-ff696756cef8

**Comment by lumixraku** [comment]
测试样张
[auto-col.json](https://github.com/user-attachments/files/17529260/auto-col.json)

File: packages/sheets/src/commands/mutations/set-worksheet-col-width.mutation.ts [hunk 5]
```diff
@@ -68,15 +108,18 @@ export const SetWorksheetColWidthMutation: IMutation<ISetWorksheetColWidthMutati
...
-
-        for (let i = 0; i < ranges.length; i++) {
-            const range = ranges[i];
-            for (let j = range.startColumn; j < range.endColumn + 1; j++) {
-                const column = manager.getColumnOrCreate(j);
-                if (typeof params.colWidth === 'number') {
-                    column.w = params.colWidth;
-                } else {
-                    column.w = params.colWidth[j] ?? defaultColumnWidth;
+        if (ranges) {
```
Comment (wzhudev) [inline packages/sheets/src/commands/mutations/set-worksheet-col-width.mutation.ts:111]: According to the definition of `ISetWorksheetColWidthMutationParams`, ranges can't be nil. So why do we have to do the `if` check here?

File: packages/sheets/src/commands/mutations/set-worksheet-col-width.mutation.ts [hunk 6]
```diff
@@ -56,6 +68,34 @@ export const SetWorksheetColWidthMutationFactory = (
...
+    worksheet: Worksheet
+): ISetWorksheetColWidthMutationParams => {
+    const { unitId, subUnitId, ranges } = params;
+    const colWidthObj: IObjectArrayPrimitiveType<Nullable<number>> = {};
+    const manager = worksheet.getColumnManager();
+
+    for (let i = 0; i < ranges.length; i++) {
+        const range = ranges[i];
+        for (let j = range.startColumn; j < range.endColumn + 1; j++) {
+            const col = manager.getColumnOrCreate(j);
```
Comment (wzhudev) [inline packages/sheets/src/commands/mutations/set-worksheet-col-width.mutation.ts:82]: Should not create?

File: packages/sheets/src/commands/mutations/set-worksheet-col-width.mutation.ts [hunk 7]
```diff
@@ -26,6 +26,18 @@ export interface ISetWorksheetColWidthMutationParams {
     colWidth: number | IObjectArrayPrimitiveType<Nullable<number>>;
 }
 
+export interface ISetWorksheetColIsAutoWidthMutationParams {
+    unitId: string;
+    subUnitId: string;
+    ranges: IRange[];
+}
+
+export interface ISetWorksheetColAutoWidthMutationParams {
```
Comment (wzhudev) [inline packages/sheets/src/commands/mutations/set-worksheet-col-width.mutation.ts:35]: These interfaces seem not used by other code so it is better to remove them.

File: packages/sheets-ui/src/controllers/render-controllers/header-resize.render-controller.ts [hunk 8]
```diff
@@ -492,19 +490,63 @@ export class HeaderResizeRenderController extends Disposable implements IRenderM
         this.disposeWithMe(
             toDisposable(
```
Comment (wzhudev) [inline packages/sheets-ui/src/controllers/render-controllers/header-resize.render-controller.ts:491]: `toDisposable` can be removed here as `disposeWithMe` can handle `Subscription`.

File: packages/sheets-ui/src/controllers/auto-width.controller.ts [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+        this.disposeWithMe(sheetInterceptorService.interceptCommand({
+            getMutations: (command: { id: string; params: ISetWorksheetColIsAutoWidthMutationParams }) => {
+                if (command.id !== SetWorksheetColIsAutoWidthCommand.id) {
+                    return {
+                        redos: [],
+                        undos: [],
+                    };
+                }
+
+                return this.getUndoRedoParamsOfColWidth(command.params.ranges);
```
Comment (wzhudev) [inline packages/sheets-ui/src/controllers/auto-width.controller.ts:104]: Should also add `unitId` and `subUnitId` as params. We shouldn't rely on the UI state if we have specific params.

File: packages/sheets/src/commands/mutations/set-worksheet-col-width.mutation.ts (hunk 7 above)
Comment (lumixraku) [reply packages/sheets/src/commands/mutations/set-worksheet-col-width.mutation.ts:35]: It used by  export interface ISetWorksheetColAutoWidthMutationParams in this file

File: packages/sheets/src/commands/mutations/set-worksheet-col-width.mutation.ts (hunk 6 above)
Comment (lumixraku) [reply packages/sheets/src/commands/mutations/set-worksheet-col-width.mutation.ts:82]: The function name "getColumnOrCreate" is not good; 

it should actually be getColumnDataWithDefaultValue.

File: packages/sheets-ui/src/controllers/auto-width.controller.ts (hunk 9 above)
Comment (lumixraku) [reply packages/sheets-ui/src/controllers/auto-width.controller.ts:104]: image
 Yes, untId & subUnitId is in return value

File: packages/sheets/src/commands/mutations/set-worksheet-col-width.mutation.ts (hunk 6 above)
Comment (wzhudev) [reply packages/sheets/src/commands/mutations/set-worksheet-col-width.mutation.ts:76]: Understood. Maybe we should change that name.

File: packages/sheets/src/commands/mutations/set-worksheet-col-width.mutation.ts [hunk 10]
```diff
@@ -56,6 +56,33 @@ export const SetWorksheetColWidthMutationFactory = (
     };
 };
 
+export const createUndoMutationsByRedos = (
```
Comment (wzhudev) [inline packages/sheets/src/commands/mutations/set-worksheet-col-width.mutation.ts:59]: This name is considered "too big" for what it actually does, especially when it is exported from the package index.ts.

**Review by wzhudev — APPROVED** [review]
LGTM

**Pull request merged**